File: src/test/java/net/flyingfishflash/loremlist/integration/domain/item/ItemDeleteTest.java
package net.flyingfishflash.loremlist.integration.domain.item;

import static net.flyingfishflash.loremlist.core.response.advice.CoreExceptionHandler.VALIDATION_FAILURE_MESSAGE;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.stream.Stream;
import net.flyingfishflash.loremlist.api.data.request.LrmItemCreateRequest;
import net.flyingfishflash.loremlist.core.response.structure.DispositionOfProblem;
import net.flyingfishflash.loremlist.core.response.structure.DispositionOfSuccess;
import net.flyingfishflash.loremlist.domain.exceptions.DomainException;
import net.flyingfishflash.loremlist.integration.domain.AbstractIntegrationTest;
import net.flyingfishflash.loremlist.integration.domain.TestData;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.TestMethodOrder;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.http.HttpMethod;

/**
 * Test methods here are order-dependent: this class is not @Transactional, and the final
 * delete-all-items test assumes the specific items already deleted by the preceding ones.
 */
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ItemDeleteTest extends AbstractIntegrationTest {

  private record ListItemRequestRecord(
      UUID listId, UUID itemId, LrmItemCreateRequest listItemCreateRequest) {}

  private List<ListItemRequestRecord> requestsAlpha;
  private List<ListItemRequestRecord> requestsBeta;

  @BeforeAll
  void setUp() throws Exception {
    purgeDomain();
    UUID listWithItems = createAndVerifyList(TestData.LIST_CREATE_REQUESTS.get(0));
    UUID listWithNoItems = createAndVerifyList(TestData.LIST_CREATE_REQUESTS.get(1));

    requestsAlpha = new ArrayList<>();
    for (var request : TestData.ITEM_CREATE_REQUESTS_ALPHA) {
      UUID itemId = createAndVerifyListItem(listWithItems, request);
      requestsAlpha.add(new ListItemRequestRecord(listWithItems, itemId, request));
    }

    requestsBeta = new ArrayList<>();
    for (var request : TestData.ITEM_CREATE_REQUESTS_BETA) {
      UUID itemId = createAndVerifyListItem(listWithNoItems, request);
      requestsBeta.add(new ListItemRequestRecord(listWithNoItems, itemId, request));
    }

    verifyRequest(HttpMethod.DELETE, "/lists/" + listWithNoItems + "/items")
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .perform();

    verifyContentValue("/lists/count", 2);
    verifyContentValue("/lists/" + listWithItems + "/items/count", requestsAlpha.size());
    verifyContentValue("/lists/" + listWithNoItems + "/items/count", 0);
    verifyContentSize("/items/with-no-lists", requestsBeta.size());
  }

  @Test
  @Order(1)
  void failsWhenInvalidUuidIsProvidedForItemId() throws Exception {
    verifyRequest(HttpMethod.DELETE, "/items/" + TestData.INVALID_UUIDS.get(0))
        .expectedDisposition(DispositionOfProblem.FAILURE)
        .statusMatcher(status().isBadRequest())
        .additionalMatchers(
            jsonPath("$.message").value(VALIDATION_FAILURE_MESSAGE + " itemId."),
            jsonPath("$.content.validationErrors.length()").value(1))
        .perform();
  }

  @Test
  @Order(2)
  void failsWhenItemToDeleteIsNotFound() throws Exception {
    UUID nonExistentId = UUID.randomUUID();
    verifyRequest(HttpMethod.DELETE, "/items/" + nonExistentId)
        .expectedDisposition(DispositionOfProblem.FAILURE)
        .statusMatcher(status().isNotFound())
        .additionalMatchers(
            jsonPath("$.content.title").value(DomainException.class.getSimpleName()),
            jsonPath("$.content.status").value("404"))
        .perform();
  }

  @ParameterizedTest(name = "fails with http 422 for ''{0}''")
  @MethodSource("requestsAlpha")
  @Order(3)
  void failsWith422(String name, ListItemRequestRecord record) throws Exception {
    verifyRequest(HttpMethod.DELETE, "/items/" + record.itemId())
        .expectedDisposition(DispositionOfProblem.FAILURE)
        .statusMatcher(status().isUnprocessableEntity())
        .perform();
  }

  Stream<Arguments> requestsAlpha() {
    return requestsAlpha.stream().map(r -> Arguments.of(r.listItemCreateRequest().getName(), r));
  }

  @ParameterizedTest(name = "succeeds for ''{0}'' [no list]")
  @MethodSource("requestsBeta")
  @Order(4)
  void succeedsNoList(String name, ListItemRequestRecord record) throws Exception {
    verifyRequest(HttpMethod.DELETE, "/items/" + record.itemId())
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .statusMatcher(status().isOk())
        .perform();
  }

  Stream<Arguments> requestsBeta() {
    return requestsBeta.stream().map(r -> Arguments.of(r.listItemCreateRequest().getName(), r));
  }

  @Test
  @Order(5)
  void deletingAllItemsSucceeds() throws Exception {
    verifyRequest(HttpMethod.DELETE, "/items")
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .perform();
    verifyContentValue("/items/count", 0);
  }
}

File: src/test/kotlin/net/flyingfishflash/loremlist/integration/domain/item/ItemDeleteTest.kt
package net.flyingfishflash.loremlist.integration.domain.item

import io.kotest.datatest.withData
import net.flyingfishflash.loremlist.api.data.request.LrmItemCreateRequest
import net.flyingfishflash.loremlist.core.response.advice.CoreExceptionHandler.Companion.VALIDATION_FAILURE_MESSAGE
import net.flyingfishflash.loremlist.core.response.structure.DispositionOfProblem
import net.flyingfishflash.loremlist.core.response.structure.DispositionOfSuccess
import net.flyingfishflash.loremlist.domain.exceptions.DomainException
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest.TestData.invalidUuids
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest.TestData.itemCreateRequestsAlpha
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest.TestData.itemCreateRequestsBeta
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest.TestData.listCreateRequests
import org.springframework.http.HttpMethod
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.util.*

class ItemDeleteTest(mockMvc: MockMvc) :
  DomainFunctionTest(mockMvc = mockMvc, body = {
    data class ListItemRequestRecord(val listId: UUID, val itemId: UUID?, val listItemCreateRequest: LrmItemCreateRequest)
    var requestsAlpha: List<ListItemRequestRecord> = emptyList()
    var requestsBeta: List<ListItemRequestRecord> = emptyList()

    beforeSpec {
      purgeDomain()
      val listWithItems = createAndVerifyList(listRequest = listCreateRequests[0])
      val listWithNoItems = createAndVerifyList(listRequest = listCreateRequests[1])

      requestsAlpha = itemCreateRequestsAlpha.map {
        ListItemRequestRecord(listId = listWithItems, itemId = createAndVerifyListItem(listWithItems, it), listItemCreateRequest = it)
      }

      requestsBeta = itemCreateRequestsBeta.map {
        ListItemRequestRecord(listId = listWithNoItems, itemId = createAndVerifyListItem(listWithNoItems, it), listItemCreateRequest = it)
      }

      performRequestAndVerifyResponse(
        method = HttpMethod.DELETE,
        instance = "/lists/$listWithNoItems/items",
        expectedDisposition = DispositionOfSuccess.SUCCESS,
      )

      verifyContentValue(url = "/lists/count", expectedValue = 2)
      verifyContentValue(url = "/lists/$listWithItems/items/count", expectedValue = requestsAlpha.size)
      verifyContentValue(url = "/lists/$listWithNoItems/items/count", expectedValue = 0)
      verifyContentSize(url = "/items/with-no-lists", expectedSize = requestsBeta.size)
    }

    describe("item deletion ") {
      context("deleting an item: /items/{item-id}") {
        it("fails when invalid uuid is provided for item id") {
          performRequestAndVerifyResponse(
            method = HttpMethod.DELETE,
            instance = "/items/${invalidUuids[0]}",
            expectedDisposition = DispositionOfProblem.FAILURE,
            statusMatcher = status().isBadRequest(),
            additionalMatchers = arrayOf(
              jsonPath("$.message").value("$VALIDATION_FAILURE_MESSAGE itemId."),
              jsonPath("$.content.validationErrors.length()").value(1),
            ),
          )
        }

        it("fails when item to delete is not found") {
          val nonExistentId = UUID.randomUUID()
          performRequestAndVerifyResponse(
            method = HttpMethod.DELETE,
            instance = "/items/$nonExistentId",
            expectedDisposition = DispositionOfProblem.FAILURE,
            statusMatcher = status().isNotFound(),
            additionalMatchers = arrayOf(
              jsonPath("$.content.title").value(DomainException::class.java.simpleName),
              jsonPath("$.content.status").value("404"),
            ),
          )
        }

        withData(
          nameFn = { "fails with http 422 for '${it.listItemCreateRequest.name}'" },
          ts = requestsAlpha,
        ) {
          performRequestAndVerifyResponse(
            instance = "/items/${it.itemId}",
            method = HttpMethod.DELETE,
            expectedDisposition = DispositionOfProblem.FAILURE,
            statusMatcher = status().isUnprocessableEntity,
          )
        }

        withData(
          nameFn = { "succeeds for '${it.listItemCreateRequest.name}' [no list]" },
          ts = requestsBeta,
        ) {
          performRequestAndVerifyResponse(
            instance = "/items/${it.itemId}",
            method = HttpMethod.DELETE,
            expectedDisposition = DispositionOfSuccess.SUCCESS,
            statusMatcher = status().isOk,
          )
        }
      }

      context("deleting all items: /items") {
        it("succeeds") {
          performRequestAndVerifyResponse(
            method = HttpMethod.DELETE,
            instance = "/items",
            expectedDisposition = DispositionOfSuccess.SUCCESS,
          )

          verifyContentValue(url = "/items/count", expectedValue = 0)
        }
      }
    }
  })
